src/Transition.js

```diff
@@ -0,0 +1,286 @@
+    super(props, context);
+
+    this.state = {
+      in: !props.in,
+      transitioning: false
+    };
+
+    this.needsTransition = true;
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.in !== this.props.in) {
+      this.needsTransition = true;
+    }
+  }
+
+  componentDidUpdate() {
+    this.processChild();
+  }
+
+  componentWillMount() {
+    this._mounted = true;
+
+    if (!this.props.transitionAppear) {
+      this.needsTransition = false;
+      this.setState({ in: this.props.in });
+    }
+  }
+
+  componentWillUnmount(){
+    this._mounted = false;
+  }
+
+  componentDidMount() {
+    if (this.props.transitionAppear) {
+      this.processChild();
+    }
+  }
+
+  processChild(){
+    let needsTransition = this.needsTransition;
+    let enter = this.props.in;
+
+    if (needsTransition) {
+      this.needsTransition = false;
+      this[enter ? 'performEnter' : 'performLeave']();
+    }
+  }
+
+  performEnter() {
+    let maybeNode = React.findDOMNode(this);
+
+    let enter = node => {
+      node = this.props.transitioningNode(node) || node;
+
+      this.props.onEnter(node);
+
+      this.safeSetState({ in: true, transitioning: true, needInitialRender: false }, ()=> {
+
+        this.props.onEntering(node);
+
+        ensureTransitionEnd(node, () => {
+          if ( this.state.in ){
+            this.safeSetState({
+              transitioning: false
+            }, () => this.props.onEntered(node));
+          }
+
+        }, this.props.duration);
+      });
+    };
+
+    if (maybeNode) {
+      enter(maybeNode);
+    }
+    else if (this.props.unmountOnExit) {
+      this._ensureNode(enter);
+    }
+  }
+
+  performLeave() {
+    let node = React.findDOMNode(this);
+
+    node = this.props.transitioningNode(node) || node;
+
+    this.props.onExit(node);
+
+    this.setState({ in: false, transitioning: true }, () => {
+      this.props.onExiting(node);
+
+      ensureTransitionEnd(node, () => {
+        if ( !this.state.in ){
+          this.safeSetState({ transitioning: false }, ()=> this.props.onExited(node));
+        }
+      }, this.props.duration);
+    });
+  }
+
+  _ensureNode(callback) {
+
+    this.setState({ needInitialRender: true }, ()=> {
+      let node = React.findDOMNode(this);
+
+      triggerBrowserReflow(node);
+
+      callback(node);
+    });
+  }
+
+  safeSetState(newState, cb){
+    if (this._mounted) {
+      this.setState(newState, cb);
+    }
+  }
+
+  render() {
+    let childProps = omit(this.props, Object.keys(Transition.propTypes).concat('children'));
+
+    let child = this.props.children;
+    let starting = this.state.needInitialRender;
+    let out = !this.state.in && !this.state.transitioning;
+
+    if ( !child || (this.props.unmountOnExit && out && !starting) ){
+      return null;
+    }
+
+    let classes = '';
+
+    // for whatever reason classnames() doesn't actually work here,
+    // maybe because they aren't always single classes?
```

Comment:
then it need to be reworded, because in that form which it is now,
it literally provokes to try and refactor :smile: 

--
and again - this is not essential for merging.